File: 1539. Kth Missing Positive Number/main.py
from typing import List


class Solution:
    def findKthPositive(self, arr: List[int], k: int) -> int:
        ARRAY_SIZE = 2001
        isMissing = [True] * ARRAY_SIZE

        for num in arr:
            isMissing[num] = False

        for i in range(1, ARRAY_SIZE):
            if not isMissing[i]:
                continue

            k -= 1
            if k <= 0:
                return i

        return -1


class Solution:
    def findKthPositive(self, arr: List[int], k: int) -> int:
        left = 0
        right = len(arr) - 1

        while left <= right:
            mid = (left + right) // 2
            if arr[mid] - (mid + 1) < k:
                left = mid + 1
            else:
                right = mid - 1

        return left + k


class Solution:
    def findKthPositive(self, arr: List[int], k: int) -> int:
        for num in arr:
            if num <= k:
                k += 1

        return k

File: 1539. Kth Missing Positive Number/main.rs
struct Solution;

const ARRAY_SIZE: usize = 2001;

impl Solution {
    pub fn find_kth_positive(arr: Vec<i32>, k: i32) -> i32 {
        let mut is_missing = vec![true; ARRAY_SIZE];

        for num in arr {
            is_missing[num as usize] = false;
        }

        let mut k = k;
        for i in 1..ARRAY_SIZE {
            if !is_missing[i] {
                continue;
            }

            k -= 1;
            if k <= 0 {
                return i as i32;
            }
        }

        return -1;
    }
}

impl Solution {
    pub fn find_kth_positive(arr: Vec<i32>, k: i32) -> i32 {
        let mut left = 0;
        let mut right = arr.len() as i32 - 1;

        while left <= right {
            let mid = (left + right) / 2;
            if arr[mid as usize] - (mid + 1) < k {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }

        left + k
    }
}

impl Solution {
    pub fn find_kth_positive(arr: Vec<i32>, k: i32) -> i32 {
        let mut k = k;

        for num in arr {
            if num <= k {
                k += 1;
            }
        }

        k
    }
}
